cloudinit/sources/DataSourceEc2.py

```diff
@@ -731,28 +735,91 @@ def convert_ec2_metadata_network_config(network_md, macs_to_nics=None,
+            # apply_network_config is False (Xenial, Bionic, Eoan)
+            if fallback_nic and fallback_nic != nic_name:
+                continue
         nic_metadata = macs_metadata.get(mac)
         if not nic_metadata:
             continue  # Not a physical nic represented in metadata
-        nic_cfg = {'type': 'physical', 'name': nic_name, 'subnets': []}
-        nic_cfg['mac_address'] = mac
-        if (nic_name == fallback_nic or nic_metadata.get('public-ipv4s') or
-                nic_metadata.get('local-ipv4s')):
-            nic_cfg['subnets'].append({'type': 'dhcp4'})
-        if nic_metadata.get('ipv6s'):
-            nic_cfg['subnets'].append({'type': 'dhcp6'})
-        netcfg['config'].append(nic_cfg)
+        dhcp_override = {'route-metric': nic_idx * 100}
+        nic_idx += 1
+        dev_config = {'dhcp4': True, 'dhcp4-overrides': dhcp_override,
+                      'dhcp6': False,
+                      'match': {'macaddress': mac.lower()},
+                      'set-name': nic_name}
+        if nic_metadata.get('ipv6s'):  # Any IPv6 addresses configured
+            dev_config['dhcp6'] = True
+            dev_config['dhcp6-overrides'] = dhcp_override
+        if apply_network_config:
+            dev_config['addresses'] = get_secondary_addresses(nic_metadata)
+            if not dev_config['addresses']:
+                dev_config.pop('addresses')  # Since we found none configured
+        else:
+            LOG.debug(
+                'Skipping secondary IP address setup.'
+                ' Datasource configure_secondary_ips is False')
+
+        netcfg['ethernets'][nic_name] = dev_config
     return netcfg
 
 
+def get_secondary_addresses(nic_metadata):
+    """Parse interface-specific nic metadata and return and any secondary IPs
+
+    :return: List of secondary IPv4 or IPv6 addresses to configure on the
+    interface
+    """
+    ipv4s = nic_metadata.get('local-ipv4s')
+    ipv6s = nic_metadata.get('ipv6s')
+    # In version < 2018-09-24 local_ipv4s or ipv6s is a str with one IP
+    secondary_ipv4 = bool(isinstance(ipv4s, list) and len(ipv4s) > 1)
+    secondary_ipv6 = bool(isinstance(ipv6s, list) and len(ipv6s) > 1)
+    if not any([secondary_ipv4, secondary_ipv6]):
+        return []
+    addresses = []
+    ip_metadata = {
+        'ipv4': {'cidr_key': 'subnet-ipv4-cidr-block',
+                 'has_secondary_ips': secondary_ipv4,
+                 'ips': ipv4s,
+                 'prefix_default': '24'},
+        'ipv6': {'cidr_key': 'subnet-ipv6-cidr-block',
+                 'has_secondary_ips': secondary_ipv6,
+                 'ips': ipv6s,
+                 'prefix_default': '128'},
+    }
+    for ip_type, md_vals in ip_metadata.items():
+        if not md_vals['has_secondary_ips']:
+            continue
+        cidr = nic_metadata.get(md_vals['cidr_key'])
+        prefix = md_vals['prefix_default']
+        if not cidr or len(cidr.split('/')) != 2:
+            LOG.warning(
```

Comment:
it would be useful to specific the nic interface name (or more likely the MAC) as this would point to a specific entry in the ec2 network metadata subtree.

Reply:
added in 408c8bfeaf7d43834896aaf4b2c46ce22a441739